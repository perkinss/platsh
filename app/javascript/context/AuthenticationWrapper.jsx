import React, { useEffect, useState, useContext } from 'react';
import User from '../model/user'

export const RootAuthContext = React.createContext();
export default ({ children }) => {
  let anyAuthVal,anyAuthData;
  try {
    anyAuthVal=JSON.parse(window.sessionStorage.getItem('authenticated'));
  }
  catch(err) {
    anyAuthVal=false;
  }
  try {
    anyAuthData = JSON.parse(window.sessionStorage.getItem('authData'))
  }
  catch(err) {
    anyAuthData = null;
  }
  const guestUser = new User({})
  const prevAuth = (anyAuthVal ? anyAuthVal : false);
  const prevAuthData = (anyAuthData ? anyAuthData : null);
  const prevUser = (anyAuthData ? new User(anyAuthData) : guestUser);
  const [authenticated, setAuthenticated] = useState(prevAuth);
  const [authData, setAuthData] = useState(prevAuthData);
  const [currentUser, setCurrentUser] = useState(prevUser);
  useEffect(
    () => {
      window.sessionStorage.setItem('authenticated', JSON.stringify(authenticated));
      window.sessionStorage.setItem('authData', (authData ? JSON.stringify(authData):null));
      setCurrentUser(authData ? new User(authData) : guestUser)
    },
    [authenticated, authData]
  );
  const defaultContext = {
    authenticated, setAuthenticated,
    authData, setAuthData,
    currentUser, setCurrentUser
  };
  return (
    <RootAuthContext.Provider value={defaultContext}>
      {children}
    </RootAuthContext.Provider>
  );
};
