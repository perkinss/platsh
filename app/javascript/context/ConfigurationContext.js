import React from 'react'

export const configuration = {
    sections: [
        {
            id: null,
            courses: [],
            name: '',
            assessment: [
                {
                    type: 'test',
                    name: '',
                    tasks: [
                    {
                        name: '',
                        available_standards: [
                            {
                                id: -1,
                                level: ''
                            }
                        ],
                        competencies: [

                        ]
                    },
                    ]
                }
            ],
            enrollees: [
                1, 2, 3, 22
            ]
        },
    ],

}


const ConfigurationContext =  React.createContext(configuration);

export { ConfigurationContext }
