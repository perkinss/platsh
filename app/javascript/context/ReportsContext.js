import React from 'react'

export const reports = {
    students: [],
    sections: [],
    selected_section: null,
    selected_student: null,
    section_overview_marks: {},
    section_competency_scores: {},
    student_section_marks: {},
    heat_map_colors: []
}


const ReportsContext =  React.createContext(reports);

export { ReportsContext }
