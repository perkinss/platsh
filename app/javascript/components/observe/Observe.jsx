import React, {useContext, useEffect, useReducer, useState} from 'react'
import {
    AppBar,
    Chip,
    Container,
    FormControl,
    Grid,
    Input,
    InputLabel,
    MenuItem,
    Select,
    Toolbar,
    Typography
} from '@material-ui/core'
import DateFnsUtils from '@date-io/date-fns';
import {useTheme, withStyles} from '@material-ui/core/styles'
import Header from '../Header'
import StudentAssessmentList from "./StudentAssessmentList"
import {ConfigurationContext} from '../../context/ConfigurationContext'
import {configReducer, initialConfigState} from "../../reducers/configurationReducer";
import CircularProgress from "@material-ui/core/CircularProgress";
import RoleValidatingWrapper from "../RoleValidatingWrapper";
import {ROLES} from "../../model/user";
import {RootAuthContext} from "../../context/AuthenticationWrapper";
import {KeyboardDatePicker, MuiPickersUtilsProvider} from '@material-ui/pickers';
import 'date-fns';

const styles = theme => ({
    root: {
        width: '100%',
        maxWidth: 360,
        backgroundColor: theme.palette.background.paper,
        position: 'relative',
        overflow: 'auto',
        maxHeight: 300,
    },
    paper: {
        padding: theme.spacing(1),
        textAlign: 'center',
        color: theme.palette.text.secondary,
        marginTop: theme.spacing(1)
    },
    toolbar: {
        display: 'flex',
        alignItems: 'center',
        justifyContent: 'flex-end',
        padding: '0 8px',
        ...theme.mixins.toolbar,
    },
    content: {
        flexGrow: 1,
        padding: theme.spacing(2),
        marginLeft: theme.spacing(1)
    },
    checkbox: {
        paddingLeft: theme.spacing(0),
        paddingRight: theme.spacing(2)
    },
    listSection: {
        backgroundColor: 'inherit',
    },
    sticky: {
        top: theme.spacing(8),
    },
});

function getStyles(name, personName, theme) {
    return {
        fontWeight:
            personName.indexOf(name) === -1
                ? theme.typography.fontWeightRegular
                : theme.typography.fontWeightMedium,
    };
}

const ITEM_HEIGHT = 48;
const ITEM_PADDING_TOP = 8;
const MenuProps = {
    PaperProps: {
        style: {
            maxHeight: ITEM_HEIGHT * 4.5 + ITEM_PADDING_TOP,
            width: 250,
        },
    },
};

function Observe (props) {
    const { classes } = props
    const { currentUser } = useContext(RootAuthContext);
    const [state, dispatch] = useReducer(configReducer, initialConfigState)

    const [sectionName, setSectionName] = useState('')
    const [filteredStudentName, setFilteredStudentName] = useState([])
    const [filteredStudentList, setFilteredStudentList] = useState([])
    const [selectedSection, setSelectedSection] = useState(null)
    const [selectedAssessment, setSelectedAssessment] = useState(null)
    const [assessmentName, setAssessmentName] = useState([])
    const [assessments, setAssessments] = useState(state.assessments || [])
    const [loadingSections, setLoadingSections] = useState(true)
    const [loadingMarks, setLoadingMarks] = useState(false)
    const [loadingCompetencies, setLoadingCompetencies] = useState(false)
    const [loadingComments, setLoadingComments] = useState(false)
    const [loadingStudents, setLoadingStudents] = useState(false)
    const [loadingAssessments, setLoadingAssessments] = useState(false)
    const [loadingAssessmentStandards, setLoadingAssessmentStandards] = useState(false)
    const [loadingAssessmentCompetencies, setLoadingAssessmentCompetencies] = useState(false)

    const theme = useTheme()

    const handleSelectSection = (event) => {
        setSectionName(event.target.value);
        const section = state.sections.find(section => section.id === event.target.value)
        setSelectedSection(section)
    }

    const handleDateChange = (date) => {
        dispatch({type: 'SET_ASSESSMENT_DATE', assessed_date: date})
    }

    const handleSelectAssessment= (event) => {
        if (assessmentName !== event.target.value) {
            setLoadingMarks(true)
            setLoadingCompetencies(true)
        }

        setAssessmentName(event.target.value)
        const assessment = state.assessments.find(assessment => assessment.id === event.target.value)
        setSelectedAssessment(assessment)
        dispatch({ type: 'SET_CURRENT_ASSESSMENT', assessment: assessment })
    }

    const assessmentList = () => {
        let students = !filteredStudentList || filteredStudentList.length < 1 ? state.students : filteredStudentList
        return (
            <Grid item xs={12} sm={12} md={12} lg={12}>
                <StudentAssessmentList
                    loadingAssessmentCompetencies={loadingAssessmentCompetencies}
                    loadingAssessmentStandards={loadingAssessmentStandards}
                    loadingMarks={loadingMarks}
                    loadingScores={loadingCompetencies}
                    loadingComments={loadingComments}
                    enrolledStudents={students}
                    selectedAssessment={selectedAssessment}
                    classes={classes}/>
            </Grid>
        )

    }

    const handleStudentNameFilter = (event) => {
        setFilteredStudentName(event.target.value)
    }

    const filterStudents = () => {
        let filteredStudents = state.students
        if (filteredStudentName.length > 0) {
            filteredStudents = state.students.filter((student) => {
                return filteredStudentName.includes(student.name)
            })
        }
        setFilteredStudentList(filteredStudents)
    }

    useEffect(() => {
        filterStudents()
    },[filteredStudentName])

    useEffect(() => {
        if (!currentUser.isTeacher()) {
            return
        }

        setLoadingAssessments(true)
        fetch('/api/assessments')
            .then(res => res.json())
            .then(
                (result) => {
                    setLoadingAssessments(false)
                    dispatch({type: 'SET_ASSESSMENTS', assessments: result})
                },
                (error) => {
                    setLoadingAssessments(false)
                    console.log("Error:", error)
                }
            )

        setLoadingSections(true)
        fetch('/api/sections')
            .then(res => res.json())
            .then(
                (result) => {
                    setLoadingSections(false)
                    dispatch({type: 'SET_SECTIONS', sections: result})
                },
                (error) => {
                    setLoadingSections(false)
                    console.log("Error:", error)
                }
            )
    },[]);

    useEffect(() => {
        if (!currentUser.isTeacher()) {
            return
        }

        if (selectedSection){
            setSelectedAssessment(null)
            setAssessmentName('')
            setFilteredStudentList([])
            setFilteredStudentName([])
            setLoadingStudents(true)
            dispatch({type: 'SET_STUDENTS', students: []})
            dispatch({type: 'SET_ASSESSMENT_DATE', assessed_date: null})
            fetch(`/api/students/enrollment/${selectedSection.id}`)
                .then(res => res.json())
                .then(
                    (result) => {
                        setLoadingStudents(false)
                        dispatch({type: 'SET_STUDENTS', students: result})
                    },
                    (error) => {
                        setLoadingStudents(false)
                        console.log("Error:", error)
                    }
                )
            const sectionAssessments  = state.assessments.filter((assessment) => {
                return assessment.sections.find((assessmentSection) => {
                    return assessmentSection.id === selectedSection.id
                })
            })

            setAssessments(sectionAssessments)
        }
    }, [selectedSection])

    useEffect(() => {
        if (!currentUser.isTeacher()) {
            return
        }

        if (selectedAssessment) {
            setLoadingAssessmentCompetencies(true)
            setLoadingAssessmentStandards(true)
            dispatch({type: 'SET_AVAILABLE_COMPETENCIES', competencies: []})
            dispatch({type: 'SET_AVAILABLE_STANDARDS', standards: []})
            dispatch({type: 'SET_ASSESSMENT_DATE', assessed_date: null})
            fetch(`api/competencies/get_by_assessment/${selectedAssessment.id}`)
                .then(res => res.json())
                .then(
                    (result) => {
                        setLoadingAssessmentCompetencies(false)
                        dispatch({type: 'SET_AVAILABLE_COMPETENCIES', competencies: result})
                    },
                    (error) => {
                        setLoadingAssessmentCompetencies(false)
                        console.log("Error:", error)
                    }
                )

            fetch(`api/standard/get_all_for_assessment/${selectedAssessment.id}`)
                .then(res => res.json())
                .then(
                    (result) => {
                        setLoadingAssessmentStandards(false)
                        dispatch({type: 'SET_AVAILABLE_STANDARDS', standards: result})
                    },
                    (error) => {
                        setLoadingAssessmentStandards(false)
                        console.log("Error:", error)
                    }
                )
            setLoadingMarks(true)
            fetch(`api/section/${selectedSection.id}/assessments/${selectedAssessment.id}/marks/list`)
                .then(res => res.json())
                .then(
                    (result) => {
                        dispatch({type: 'SET_MARKS_FROM_FETCHED_OBSERVATIONS', observations: result})
                        dispatch({type: 'SET_ASSESSMENT_DATE', assessed_date: result.assessed_date})
                        setLoadingMarks(false)
                    },
                    (error) => {
                        console.log("Error:", error)
                        setLoadingMarks(false)
                    }
                )
            setLoadingCompetencies(true)
            fetch(`api/section/${selectedSection.id}/assessments/${selectedAssessment.id}/scores/list`)
                .then(res => res.json())
                .then(
                    (result) => {
                        dispatch({type: 'SET_SCORES_FROM_FETCHED_OBSERVATIONS', observations: result})
                        setLoadingCompetencies(false)
                    },
                    (error) => {
                        console.log("Error:", error)
                        setLoadingCompetencies(false)
                    }
                )
            setLoadingComments(true)
            fetch(`api/comments/${selectedAssessment.id}`)
                .then(res => res.json())
                .then(
                    (result) => {
                        // collect comments into
                        // {studentId: {taskId: comment, taskId: comment}

                        dispatch({type: 'SET_COMMENTS_FROM_FETCH', comments: result})
                        setLoadingComments(false)
                    },
                    (error) => {
                        console.log("Error:", error)
                        setLoadingComments(false)
                    }
                )
        }

    }, [selectedAssessment])

    return (
        <React.Fragment>

            <ConfigurationContext.Provider value={{dispatch: dispatch, state: state}} >

            <Header currentPage={"Observe"}>
                <Container maxWidth={'xl'}>
                        <div className={classes.toolbar} />
                            <RoleValidatingWrapper allowedRoles={[ROLES.TEACHER]}>
                                <AppBar position="sticky" className={classes.sticky}>
                                    <Toolbar>
                                        <Grid container spacing={theme.spacing(1)} >
                                            <> { loadingSections || loadingAssessments ?
                                                <Grid item xs={12} sm={12} md={6} lg={4}>
                                                    <CircularProgress color={"secondary"} />
                                                </Grid>
                                                :
                                                <Grid item xs={12} sm={4} md={3} lg={3}>
                                                    <FormControl className={classes.formControl} fullWidth={true}>
                                                        <InputLabel htmlFor="select-section">Section</InputLabel>
                                                        <Select
                                                            value={sectionName}
                                                            onChange={handleSelectSection}
                                                        >
                                                            {state.sections.map(section => (
                                                                <MenuItem key={section.id} value={section.id}>
                                                                    {section.name}
                                                                </MenuItem>
                                                            ))}
                                                        </Select>
                                                    </FormControl>
                                                </Grid>
                                            }
                                            </>
                                            <>{
                                                selectedSection ?

                                                    <>{ assessments.length === 0 ?
                                                        <Grid item xs={12} sm={12} md={12} lg={12} style={{paddingTop: 0}}>
                                                            <Typography variant={'body2'}>Hint: You need to add some assessments to this section.</Typography>
                                                        </Grid>
                                                        :
                                                        <Grid item xs={12} sm={4} md={3} lg={3}>
                                                            <FormControl className={classes.formControl} fullWidth={true}>
                                                                <InputLabel htmlFor="select-asmt">Assessment:</InputLabel>
                                                                <Select
                                                                    value={assessmentName || ""}
                                                                    onChange={handleSelectAssessment}
                                                                    input={<Input id="select-asmt"/>}
                                                                >
                                                                    {assessments.map(assessment => (
                                                                        <MenuItem key={assessment.id} value={assessment.id}>
                                                                            {assessment.name}
                                                                        </MenuItem>
                                                                    ))}
                                                                </Select>
                                                            </FormControl>

                                                        </Grid>
                                                    }
                                                    </> : ''
                                            }</>
                                            <>
                                                { selectedSection && !loadingAssessments && assessments.length > 0 ?
                                                    <>{ loadingStudents ?
                                                        <Grid item xs={12} sm={6} md={6} lg={4}>
                                                            <CircularProgress color={"secondary"}/>
                                                        </Grid>
                                                        :
                                                        <>
                                                            { state.students.length === 0
                                                                ?
                                                                <Grid item xs={12} sm={12} md={12} lg={12} style={{paddingTop: 0}}>
                                                                    <Typography variant={'body2'}>Hint: You need to add some students to this
                                                                        section before you can assess them.</Typography>
                                                                </Grid>
                                                                :
                                                                <Grid item xs={12} sm={4} md={3} lg={3}>
                                                                    <FormControl className={classes.formControl} fullWidth={true}>
                                                                        <InputLabel htmlFor="select-multiple-chip">Filter Students</InputLabel>
                                                                        <Select
                                                                            multiple
                                                                            value={filteredStudentName}
                                                                            onChange={handleStudentNameFilter}
                                                                            input={<Input id="select-multiple-chip"/>}
                                                                            renderValue={selected => (
                                                                                <div className={classes.chips}>
                                                                                    {selected.map(value => (
                                                                                        <Chip key={value} label={value}
                                                                                              className={classes.chip}/>
                                                                                    ))}
                                                                                </div>
                                                                            )}
                                                                            MenuProps={MenuProps}
                                                                        >
                                                                            {state.students.map(student => (
                                                                                <MenuItem key={student.id} value={student.name}
                                                                                          style={getStyles(student.name, filteredStudentName, theme)}>
                                                                                    {student.name}
                                                                                </MenuItem>
                                                                            ))}
                                                                        </Select>
                                                                    </FormControl>
                                                                </Grid>
                                                            }</>
                                                    }</>
                                                    : ''}</>
                                            <>
                                                {selectedSection && !loadingAssessments && assessments.length > 0 ?
                                                    <>{loadingStudents ?
                                                        <Grid item xs={12} sm={4} md={3} lg={2}>
                                                            <CircularProgress color={"secondary"}/>
                                                        </Grid>
                                                        :
                                                        <>
                                                            {state.students.length === 0
                                                                ?
                                                                ''
                                                                :
                                                                <Grid item xs={12} sm={4} md={3} lg={3}>
                                                                    <MuiPickersUtilsProvider classes={classes.formControl} utils={DateFnsUtils}>
                                                                        <KeyboardDatePicker
                                                                            disableToolbar
                                                                            variant={'inline'}
                                                                            format={'MMMM dd, yyyy'}
                                                                            margin={'normal'}
                                                                            fullWidth={true}
                                                                            style={{marginTop: '0'}}
                                                                            label={'Assessment Date'}
                                                                            value={state.assessed_date}
                                                                            onChange={handleDateChange}
                                                                            KeyboardButtonProps={{
                                                                                'aria-label' : 'assessment date',
                                                                            }}
                                                                        />
                                                                    </MuiPickersUtilsProvider>
                                                                </Grid>
                                                            }
                                                        </>
                                                    }</>
                                                    :
                                                    ''
                                                }
                                                </>
                                        </Grid>
                                    </Toolbar>
                                </AppBar>

                                {selectedAssessment && state.student_marks && state.comments && state.student_competencies && assessmentList()}
                            </RoleValidatingWrapper>
                </Container>
            </Header>
            </ConfigurationContext.Provider>
        </React.Fragment>

    );

}
export default withStyles(styles)(Observe);