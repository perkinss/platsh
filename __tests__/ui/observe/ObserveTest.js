import React from 'react'
import Observe from '../../../app/javascript/components/observe/Observe'
import { createShallow, createRender } from '@material-ui/core/test-utils';
import Enzyme from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import expect from 'expect'
import {ConfigurationContext} from "../../../app/javascript/context/ConfigurationContext";
import {MemoryRouter} from "react-router";
import AuthenticationWrapper from "../../../app/javascript/context/AuthenticationWrapper";

Enzyme.configure({ adapter: new Adapter() });

describe('<Observe />', () => {
    let shallow, render;


    let state = {
        available_competencies : [{id: 1, description: "Competency 1"},{id: 2, description: "C 2"},{id: 3, description: "C 3"}],
        students: [{id: 7, name: 'Jane'}],
        selected_section: {name: 'aha'},
        selected_assessment: {
            tasks: [{id: 15, competencies: [1,2,3]}]
        },
        student_competencies: {7: {15: {1: 4, 2: 3, 3: 0}}}
    }

    beforeAll(() => {
        shallow = createShallow();
        render = createRender()
    })

    it('should match the snapshot', () => {
        const wrapper = shallow(<Observe />);
        expect(wrapper).not.toBeNull()
        expect(wrapper).toMatchSnapshot()
    });

    it('should have circular progress when loading sections', () => {
        window.sessionStorage.setItem('authenticated', 'true')
        window.sessionStorage.setItem('authData', JSON.stringify({email:'placeholder',roles:['teacher']}))
        const wrapper = render(
            <MemoryRouter>
                <AuthenticationWrapper>
                    <ConfigurationContext.Provider value={{dispatch: () => {}, state: state}} >
                        <Observe />
                    </ConfigurationContext.Provider>
                </AuthenticationWrapper>
            </MemoryRouter>);
        expect(wrapper.html()).toContain("Observe")
        expect(wrapper.html()).toContain("CircularProgress")
    });


})
