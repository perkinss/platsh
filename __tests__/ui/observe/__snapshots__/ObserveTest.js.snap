// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Observe /> should match the snapshot 1`] = `ShallowWrapper {}`;
